NIFI-12880 Add processor DeleteFile

EndzeitBegins opened the pull request:
<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary

Adds new processor `DeleteFile` to the bundle standard-processors.

See issue [NIFI-12880](https://issues.apache.org/jira/browse/NIFI-12880) for the motivation behind the addition.

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### Licensing

- [x] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [x] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

EndzeitBegins commented:
Sadly `nifi-web-frontend` failed. I'm not quite sure why that is but I'm certain that's not due to the changes in the PR. 🤷🏻‍♂️

Would you be so kind to start the failed jobs once more @ChrisSamo632?

EndzeitBegins commented:
Thank you @ChrisSamo632. 👍🏻

EndzeitBegins commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DeleteFile.java:
```diff
@@ -0,0 +1,138 @@
+
+            Files.delete(filePath);
+
+            session.transfer(flowFile, REL_SUCCESS);
```
That sounds very reasonable. I just looked for reference in other processors and `DeleteS3Object` is [doing exactly that](https://github.com/apache/nifi/blob/90c7dba34f08921b0b3b528c6a9fc20bf298a288/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/DeleteS3Object.java#L136).

mattyb149 replied:
That's because DeleteS3Object is deleting from an external storage system, not sure if that's appropriate for the local filesystem or not.

joewitt replied:
It seems appropriate as 'remote' here mostly just indicates 'it invoked something to happen outside of NiFi itself.

EndzeitBegins replied:
That's right. However, one could argue that the file system is external to the NiFi flow itself. 
I think it's the most fitting of all existing provenance event types.
And as this is a destructive action a provenance record seems very sensible.

joewitt commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DeleteFile.java:
```diff
@@ -0,0 +1,138 @@
+import java.util.Set;
+
+@SupportsBatching(defaultDuration = DefaultRunDuration.TWENTY_FIVE_MILLIS)
+@InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
```
Unless a clear intent/use case already exists please keep this input required as that satisfies the specific use case intended.

EndzeitBegins replied:
That's an interesting remark. 

The other existing `DeleteX` processors I looked into all required an input FlowFile, so I think that's a reasonable default to start with.
Especially, since using it with a `GenerateFlowFile` is an easy workaround. 

For now, I would argue that `INPUT_REQUIRED` is a sensible start. 
Allowing to use the processor without input should be relatively easy to support in a later point in time, in case that should be deemed useful.
However, the opposite is not true I assume

joewitt commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DeleteFile.java:
```diff
@@ -0,0 +1,138 @@
+            .addValidator(StandardValidators.createDirectoryExistsValidator(true, false))
+            .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+            .build();
+    public static final PropertyDescriptor FILENAME = new PropertyDescriptor.Builder()
```
DeleteFile is the correct name either way.

joewitt replied:
This should be designed to delete a single specified file and not be recursing/etc..

mattyb149 replied:
Also I think the common convention for paths and files is to use `identifiesExternalResource()`, that might even alleviate the need for a validator.

EndzeitBegins replied:
I looked into `identifiesExternalResource()` a little more. To my knowledge it is rather used to reference a (more or less static) resource, such as a configuration file on the NiFi node.

However, the use case here feels somewhat different. The directory and file are not a external resource required for the processor to work but rather the actual focus of the process.

The other `*File` processors such as `ListFile` or `PutFile` do NOT make use of `identifiesExternalResource()` either. Unless I receive additional feedback on this, I would stuck using the existing `StandardValidators.createDirectoryExistsValidator` as it's used in for example in `ListFile` for now.

jrsteinebrey commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DeleteFile.java:
```diff
@@ -0,0 +1,138 @@
+            getLogger().error("Failed to delete file '{}' in directory '{}'", filename, directoryPathProperty, ioException);
+
+            session.penalize(flowFile);
+            session.transfer(flowFile, REL_FAILURE);
```
For each place that transfers to REL_FAILURE, I think it would be helpful to add to the flowFile a writesAttribute with an error message about why it failed to help humans troubleshooting a failure. For this case of REL_FAILURE I would use wording like the log error on line 132 and add the exception class name and the exception,getMessage(). I suggest writing an error attribute for line 122 above also.

joewitt replied:
@jrsteinebrey  We should not start having a standard of writing exception class names in flowfile attributes nor should we make it a standard to have raw exception messages in an attribute.  

I certainly agree we can and should do more to provide more context to the user but doing it this way is not nearly as effective as the component/extension developer being thoughtful about which exceptions can occur and what would be best to convey to the user both in terms of what we log, what we may add as context in flowfile attributes, and what relatioships we give them to make different flow design decisions.

jrsteinebrey replied:
@joewitt @EndzeitBegins 
OK, Joe. I retract my suggestion based on your points.

EndzeitBegins commented:
I've added a commit that:
1. upon success, emits a provenance event of type `REMOTE_INVOCATION` 
2. upon failure, adds failure details to the FlowFile's attributes

I'd appreciate another review.

EndzeitBegins commented:
The frontend build failed again without related changes. :( Would you be so kind to kick off the CI process once more @ChrisSamo632?

exceptionfactory commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DeleteFile.java:
```diff
@@ -0,0 +1,174 @@
+                attribute = DeleteFile.ATTRIBUTE_EXCEPTION_MESSAGE,
+                description = "The message of the exception thrown during processor execution. Only available if an exception caused the FlowFile to be routed to relationship 'failure'.")
+})
+public class DeleteFile extends AbstractProcessor {
```
This Processor needs the `@Restricted` annotation and should require the `READ_FILESYSTEM` and `WRITE_FILESYSTEM` permissions.

exceptionfactory reviewed (changes requested):
@EndzeitBegins I have not reviewed the details of this Processor, but one important note is that it should include required permissions for filesystem access.

EndzeitBegins commented:
Thank you for your feedback @exceptionfactory.

As there has been quite some work on `main` since I've opened the PR, I've rebased the PR on the latest commits of `main`.
Also I added a commit that adds the `@Restriction` annotations as per your feedback.

exceptionfactory reviewed (approved):
Thanks for making the adjustments @EndzeitBegins. The `displayName()` attribute can be removed from the property descriptor since it matches the `name()`, but I will handle that when merging.
